oqsprov/oqsprov.c

```diff
@@ -739,21 +750,31 @@ int OQS_PROVIDER_ENTRYPOINT_NAME(const OSSL_CORE_HANDLE *handle,
             ERR_raise(ERR_LIB_USER, OQSPROV_R_OBJ_CREATE_ERR);
             fprintf(stderr, "error registering NID for %s\n",
                     oqs_oid_alg_list[i + 1]);
-            return 0;
+            goto end_init;
+        }
+
+        /* create object (NID) again to avoid setup corner case problems
+         * see https://github.com/openssl/openssl/discussions/21903
+         * Not testing for errors is intentional.
+         * At least one core version hangs up; so don't do this there:
+         */
+        if (strcmp("3.1.0", ossl_versionp)) {
```

Comment:
Is 3.1.0 the only version that hangs here? The comment above "_At_ _least_ _one_ core version hangs up" seems to suggest that there are more versions where this happens.

Reply:
I didn't test all versions. If you have a fast machine that permits that, please do (and amend the code suitably).

Reply:
I'll do some test runs with other versions.

Reply:
I ran tests for all 3.x.x release versions, 3.1.0 was the only version that hangs.

Unrelated to this, tests with versions 3.0.0, 3.0.4 and 3.0.10 fail (and 3.0.1 tests skip as expected in with runtests.sh). @baentsch do you want me to amend this PR to skip 3.0.0 and 3.0.4 tests, and fix the grep that causes 3.0.10 to be mistakenly skipped?

3.0.0
```
...
OpenSSL 3.0.0 7 sep 2021 (Library: OpenSSL 3.0.0 7 sep 2021)
error registering dilithium2 with no hash
error registering dilithium2 with no hash
Providers:
  default
    name: OpenSSL Default Provider
    version: 3.0.0
    status: active
    build info: 3.0.0
    gettable provider parameters:
      name: pointer to a UTF8 encoded string (arbitrary size)
      version: pointer to a UTF8 encoded string (arbitrary size)
      buildinfo: pointer to a UTF8 encoded string (arbitrary size)
      status: integer (arbitrary size)
error registering dilithium2 with no hash
oqsprovider not registered. Exit test.
```

3.0.4
```
...
The following tests FAILED:
    3 - oqs_groups (Subprocess aborted)
Errors while running CTest
Output from these tests are in: /data/bhe/git/oqs-provider/_build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.

Tests failed.
```

3.0.10
```
...
Test setup:
LD_LIBRARY_PATH=/data/bhe/git/oqs-provider/.local/lib64
OPENSSL_APP=/data/bhe/git/oqs-provider/openssl/apps/openssl
OPENSSL_CONF=/data/bhe/git/oqs-provider/scripts/openssl-ca.cnf
OPENSSL_MODULES=/data/bhe/git/oqs-provider/_build/lib
Version information:
OpenSSL 3.0.10 1 Aug 2023 (Library: OpenSSL 3.0.10 1 Aug 2023)
Skipping testing of buggy OpenSSL 3.0.1
```

Reply:
> @baentsch do you want me to amend this PR to skip 3.0.0 and 3.0.4 tests, and fix the grep that causes 3.0.10 to be mistakenly skipped?

As those things are truly unrelated to this topic, a separate PR seems better.